Fix #39: README issues w/ Docker Permissions

tarfeef101 opened the pull request:
- don't tell users to run as root
- use compose if we're going to use compose
- simplify commands for up/down/stop

Would be great if we could have a default `docker-compose` file that wouldn't require specifying the file each time, but not necessary. Would also be nice to not use a ubuntu base image, they're rather bloated. Would be even better to separate runtime and building images for faster builds and smaller images overall. I'd also like to not make the interactive container manually, but simply define a runtime container as a service in compose, and users could run `docker-compose run wallet`, for example, and that'd dump them into a shell within the environment specified (and since it's in the same compose file, the network is automatically correct).

If any/all of these additional changes are desired, that can be added to this or made as separate PRs too.

closes #39

wadagso-gertjaap committed:
Add package net-tools, which contains netstat

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

wadagso-gertjaap committed:
Add tty:true to prevent stdout buffering

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

wadagso-gertjaap committed:
Use CMD-SHELL for healthcheck to enable piping

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Refactor trivial manual serialization to use helper methods with error detection

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Add helper method for serializing into a shared buffer

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Remove trivial direct uses of buffer_serializer requiring a shared_ptr

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Rename atomizer messages source files to format

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Move atomizer messages to their own file

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Refactor atomizer state machine to use structs and helpers for serialization

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Calculate size of object and allocate before serialization

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Support multiple consumers from blocking queue

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement serialization for cbdc::buffer

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Expose parse_ip_port function from config namespace

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Implement siphash algorithm for cbdc::buffer

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

HalosGhost committed:
Initial Commit

Co-authored-by: James Lovejoy <jameslovejoy1@gmail.com>
Co-authored-by: Spencer Connaughton <spencer@spencerconnaughton.com>
Co-authored-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>
Co-authored-by: Cory Fields <cory-nospam-@coryfields.com>
Co-authored-by: Neha Narula <narula@gmail.com>
Co-authored-by: Sam Stuewe <stuewe@mit.edu>
Co-authored-by: Kevin Karwaski <kevin.karwaski@gmail.com>
Co-authored-by: Viktor Urvantsev <israelipaperboy@outlook.com>
Co-authored-by: jallen-frb <Jonathan.Allen@bos.frb.org>
Co-authored-by: Anders Brownworth <anders-github@evantide.com>

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Move architectures into their own directory based on transaction format

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

metalicjames committed:
Retry building the raft cluster forever

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Reduce watchtower block cache depth and increase target block time

When using the docker compose configuration for the atomizer architecture, the default watchtower block cache depth results in docker killing the watchtower executable. This prevents the compose stack from starting successfully. To reduce the memory allocation in the watchtower, reduce the block cache depth to 50 blocks. Also increase the target block time to 3 seconds. This gives 150s before blocks are evicted from the cache, which should be plenty of time to run sync from the client.

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Wait for coordinator response before returning from client

The 2PC client should wait for the coordinator response to a mint operation. Otherwise, the client will exit before receiving the response meaning a manual sync is required to confirm the mint transaction. This commit adds a 5 second timeout before returning from send_mint_tx.

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Add a config flag to select whether to wait for followers when building a raft cluster

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

Commit:
docs: update README's docker instructions:

- don't tell users to run as root
- use compose if we're going to use compose
- simplify commands for up/down/stop

Signed-off-by: tarfeef101 <tarfeef101@tareef.tech>

HalosGhost commented on README.md:
```diff
@@ -55,26 +55,29 @@ The easiest way to compile the code and run the system locally is using [Docker]
 
-Don't forget to run the docker daemon!
+### Recommended post-install steps
+- Enable the Docker daemon as detailed [here](https://docs.docker.com/engine/install/linux-postinstall/#configure-docker-to-start-on-boot)
```
```suggestion
- [Enable the Docker daemon](https://docs.docker.com/engine/install/linux-postinstall/#configure-docker-to-start-on-boot)
```

Bit of a nit, but when possible, I like to follow  the [WCAG Guidelines for links](https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html) and use more descriptive link text.

HalosGhost commented on README.md:
```diff
@@ -55,26 +55,29 @@ The easiest way to compile the code and run the system locally is using [Docker]
+**Note:** You will need to both run the system and interact with it. To do this, you should run the system in **detached** mode (in the background), and then you can launch a second container in which to run commands. This is detailed below.
 
-**Note:** You will need to both run the system and interact with it; you can either use two shells, or you can add the `--detach` flag when launching the system (note that it will then remain running till you stop it, e.g., with `docker stop`).
 Additionally, you can start the atomizer architecture by passing `--file docker-compose-atomizer.yml` instead.
```
This is a more general question: do you think it makes more sense for us to expand all flags that we mention (rather than using their short-option form) for clearer documentation?

If not, we should at least probably be consistent throughout the document so we're not referring to an option by its long and short form in different places.

tarfeef101 replied:
I would lean towards short-forms to encourage people to use things more efficiently, and where it is difficult to infer the meaning, to simply explain in the instruction relating to the command snippet.

HalosGhost replied:
I don't think that's unreasonable; can you walk through the rest of the README then and make sure that we refer to each option consistently?

HalosGhost commented on README.md:
```diff
@@ -55,26 +55,29 @@ The easiest way to compile the code and run the system locally is using [Docker]
+### Recommended post-install steps
+- Enable the Docker daemon as detailed [here](https://docs.docker.com/engine/install/linux-postinstall/#configure-docker-to-start-on-boot)
+  - It is recommended to read the other instructions here to customize other features you may find useful
+- Follow [these instructions](https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user) to enable the ability to run docker as a non-root user
```
I'm trying to figure out if there's a sensible way to rephrase this to also have a more descriptive link text, but nothing's immediately coming to mind for me…

tarfeef101 replied:
changed it a bit

HalosGhost reviewed (changes requested):
This looks like a solid improvement!

I have a couple of small changes requested, and there are a few logistical things that should be handled as well:

* In accordance with [the Contribution Guide](https://github.com/mit-dci/opencbdc-tx/blob/trunk/docs/contributing.md#intellectual-property), this commit needs a DCO to be considered for inclusion. (we're going to be adding a CI task to check this automatically soon so that this gets handled quickly and easily, sorry it's not in-place yet!)
* It would be great if the commit message header could also be more descriptive (e.g., `docs: improve README's docker walkthrough`)

HalosGhost commented:
> Would be great if we could have a default `docker-compose` file that wouldn't require specifying the file each time, but not necessary.

I think we're unlikely to go this route (because we want it to be as simple as possible to run each architecture, and not suggest that either one is official or default).

> Would also be nice to not use a ubuntu base image, they're rather bloated. 

This sounds interesting!
In part, we used ubuntu because it gives a simple target for people to use to know that a bare machine will work to run the code, but I'm definitely open to seeing alternatives!

> Would be even better to separate runtime and building images for faster builds and smaller images overall.

This is also interesting!
I'd love to see an example or hear more about how this would work (anything we can do to improve the code's usability is on-the-table).

> I'd also like to not make the interactive container manually, but simply define a runtime container as a service in compose, and users could run `docker-compose run wallet`, for example, and that'd dump them into a shell within the environment specified (and since it's in the same compose file, the network is automatically correct).
That sounds terrific.
I'd love to see a separate PR for that (and for several of the things above).

Thank you for pitching in a hand; welcome to the project!

HalosGhost committed:
docs: Correct statistics in README

The statistics in the README are outdated, and do not refer to the
latest demonstrated numbers in the technical paper.

This minor change corrects the statistics to match the paper
appropriately.

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

theuni committed:
build: Check for buildtime support of cpu features

Signed-off-by: Cory Fields <cory-nospam-@coryfields.com>

wadagso-gertjaap committed:
Align NuRaft version Dockerfile with configure.sh

Signed-off-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>

wadagso-gertjaap committed:
Run make commands in Docker on all cores

Signed-off-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>

wadagso-gertjaap committed:
Add missing cbdc::atomizer::request

Signed-off-by: Gert-Jaap Glasbergen <gertjaap@wadagso.com>

jsoref committed:
Spelling fixes for various typos

* asynchronously
* attestation
* committed
* corresponding
* entirety
* environment
* forward
* inordinate
* linearizability
* nonexistent
* occurred
* preceded
* program
* receive
* synchronous
* the
* unknown
* wellformed

Signed-off-by: Josh Soref <jsoref@users.noreply.github.com>

HalosGhost committed:
docs: replace a few links in Contribution Guide

In particular, we were made aware that one of the resources we linked
to included significant false information.

Instead, we have replaced these links to a generic project which
tracks CBDC projects worldwide.

Signed-off-by: Sam Stuewe <stuewe@mit.edu>

dmtrinh committed:
Removed extraneous \return command for Doxygen

Signed-off-by: Duc M. Trinh <duc.m.trinh@gmail.com>

tarfeef101 commented:
> -- Snip --

re: more advanced dockering:

### Different base image
The de-facto standard of alpine for small/fast images is what I'd tend toward, I can make another PR using that instead. The biggest thing people run into there is that fact that it uses `musl` instead of `glibc` by default, but that may not actually be a problem, and if it is, it can be ameliorated, so that can be tested out fairly easily.

### Runtime vs Building images
Basically, we could do the following:
- define an image to build the executable(s). it will have all the dependencies to build the output program, and be layered in a way such that it can be built quickly, even if slightly larger, as compute time is generally more valuable than storage. For example:
```
# alpine:3.15.0
FROM alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3
RUN install some deps
COPY some_infrequently_changed_files
RUN first_build_command
COPY frequently_changed_files
RUN second_build_command
```
- define a second runtime images with only runtime dependencies, and then the executable built by the builder image. For example:
```
# get binary from builder image
FROM previous_image:tag as builder
FROM alpine:latest
RUN install_runtime_deps
COPY --from=builder /opt/binary /opt
ENTRYPOINT ["/bin/sh", "-c", "/opt/binary", "--option", "value"]
```

This would be done so the runtime image can be distributed separately from the builder image, and be kept much smaller (it only needs runtime dependencies, and the binaries. In addition, the builder image would be smaller (for the same reasons, but only needing build dependencies), and could be built faster as a result.

### Wallet vs System Services
By leveraging the previous suggestion, you could have 2 different services laid out like this:
```
services:
  wallet:
    build: src/wallet/Dockerfile
    entrypoint: ["/bin/bash"]
    command: ["/opt/client-cli", "--relevant-option"]
  system:
    build: src/system/Dockerfile
    command: ["system_executable", "--relevant-option"]
```
Then you can run `docker-compose up -d system` (and add more services if more are required), and `docker-compose run wallet` will dump you right into a shell, and optionally (as shown above) in an interactive session within the executable if that's supported (if not just nix the `command` line and you'll just have a regular bash shell).

HalosGhost commented on README.md:
```diff
@@ -55,26 +55,29 @@ The easiest way to compile the code and run the system locally is using [Docker]
-$ cd opencbdc-tx                 # change to the project directory
-$ sudo -s                        # open a root shell (needed for docker)
-# docker build . -t opencbdc-tx  # build the container
+# docker-compose -f docker-compose-2pc.yml build
```
Here, and below, these commands are intended to be run as non-root, so they should be replaced with a `$` (Note though that the `#` inside the container is still correct and should be left as-is).

HalosGhost replied:
I'm also wondering if we should call out that despite this command using the `docker-compose-2pc.yml` file, it's building the full image (so you don't need to rebuild to run the atomizer architecture). Thoughts?

tarfeef101 replied:
I think that's worthwhile, one wouldn't intuitively expect multiple compose files to all use the same image, so that sounds like a good improvement.

HalosGhost commented:
> The de-facto standard of alpine for small/fast images is what I'd tend toward, I can make another PR using that instead. The biggest thing people run into there is that fact that it uses `musl` instead of `glibc` by default, but that may not actually be a problem, and if it is, it can be ameliorated, so that can be tested out fairly easily.

I was wondering if that's what you had in-mind. I like Alpine, and like the idea of smaller images, but I'm a little concerned that moving this direction might confuse the situation with `./scripts/configure.sh`. @metalicjames where do you come down on this?

> ### Runtime vs Building images

I think this could be really interesting. It offers a reasonable route to distributing images for people that just want to play around with the system (the runtime images). I do wonder though, code changes will necessitate rebuilding both images, right? How likely is it that you'd want to rebuild the runtime images without also rebuilding the builder images?

> ### Wallet vs System Services

This is also still very interesting to me, but I'm wondering where the line gets drawn. We have two separate compose files right now, at what point would it make more sense for this solution to define a service for each architecture, and combine the files? My immediate concern there is that it'd get harder to see the separation between the two architectures. What do you think about this?

metalicjames commented:
> where do you come down on this?

We don't use the configure script for the docker images right now, so switching to Alpine images is probably a good idea.

I'm a fan of all the proposals above, but they should be split out into separate PRs to keep this focused on the readme changes.

HalosGhost commented:
> We don't use the configure script for the docker images right now, so switching to Alpine images is probably a good idea.

I'm willing to accept that this may be an example of my being overly-cautious. ☺

> they should be split out into separate PRs to keep this focused on the readme changes.

Sounds perfect.

dmtrinh committed:
Disable RTTI for all env. Fix macOS builds.

Signed-off-by: Duc M. Trinh <duc.m.trinh@gmail.com>

cdesch committed:
Adding Dep and  docs for running the test suite in docker

- Adding process for running unit tests
- Add `lcov` to Dockerfile for testing with `./scripts/test.sh`. `lcov` is needed for code coverage as noted in the [Contributing Docs](https://github.com/mit-dci/opencbdc-tx/blob/trunk/docs/contributing.md#testing)

Signed-off-by: Chris Desch <cdesch@gmail.com>

HalosGhost commented:
Note that we merged a small additional walkthrough to the README that should probably be updated in this PR.

metalicjames committed:
Simplify shard seeder tool to use config file

Also unify output filenames so shards and shard seeder use the same names.

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

metalicjames committed:
Update default config values to use fixed size transactions for benchmarking

Signed-off-by: James Lovejoy <jameslovejoy1@gmail.com>

Commit:
Merge remote-tracking branch 'upstream/trunk' into trunk
sorry, been a busy week, I'll try and make the couple changes here in
the next day or two, and then I can tackle those larger docker changes

Signed-off-by: tarfeef101 <tareef@tareef.tech>

HalosGhost commented:
@tarfeef101 following the correction mentioned in the NEWS post, this PR needs to be rebased.
